README.md
# mdReaderJS

Application javascript de lecture d'une fiche de métadonnées conforme ISO-19139/INSPIRE.
Cette application constitue l'un des modules du projet mdViewer

## Projet mdViewer

Le projet "mdViewer" vise à proposer une solution simple de consultation de fiches de métadonnées issues de services web (CSW) ou de fichiers XML stockés en ligne (Iso 19139) en utilisant une logique client/serveur.
Il n'a pas pour objectif de remplacer des solutions complètes tel que GéoSources ou GéoNetwork.

L'application se compose de 3 modules indépendants:

  - cswReaderJS : permet de lire un flux afficher la liste des fiches de métadonnées de façon synthétique.
  - mdReaderJS : permet de lire une fiche de métadonnées et l'afficher de façon complète selon le profil CIGAL.
  - xml2csw : permet de simuler un serveur csw minimaliste à partir d'une liste de fichiers XML pour pouvoir les consulter via cswReaderJS et mdReaderJS.

Cette application s'inspire des travaux réalisé pour le sViewer développé par Géobretagne.

## Technologie:

Le module cswReaderJS est développé principalement via du JavaScript, du HTML, du CSS 3. Il s'apuie notamment sur les bibliothèques suivantes:

JQuery pour l'interaction javacript
Mustache et Mustache.js comme système de template pour la mise en forme des pages HTML
Uikit pour la présentation des pages et le rendu
Un script PHP est utilisé pour appeler les pages distantes et permettre de réaliser des requêtes AJAX "cross-domain".
Il peut être facilement remplacé par un script dans un autre langage comme Python, Java ou autre (développement à prévoir).

## Principes:

A partir d'une URL vers un serveur CSW et un identifiant de fiche transmis en paramètre, le module génère une requête de type GetRecordByID et affiche la métadonnée envoyée en retour selon le modèle spécifié.

## Installation:

  - Télécharger le fichier zip contenant les sources.
  - Dézipper le fichier téléchargé sur le serveur
  - L'application est fonctionnelle.

## Paramétrage:

L'url est de la forme: http://www.mondomaine.ext/chemein/vers/mdReader/?url=http://url/du/service/csw&id=identifiant_de_la_fiche&tpl=template.html

Les paramètres sont:
  - "url" = URL du serveur CSW
  - "id" = Identifiant de la fiche de métadonnées appelée par getRecordById
  - "tpl" = Nom du fichier de template (localisé dans le dssier "templates" de l'application. Le paramétrage du template par défaut s'effectue dans le fichier "js/mdReader.js".

## Démonstration:

http://www.cigalsace.net/mdReaderJS/0.05/index.html?url=https://www.cigalsace.org/geonetwork/srv/fre/csw-geocatalogue&id=FR-236700019-ORTHO-PIR-20112012-CIGAL&tpl=bootstrap.html


history.txt
# mdReaderJS

## TODO :
    - ajouter les variables de label dans le template bootstrap.html
    - ajouter parametre "lang" dans l'URL et affichier labels en fonction

## v. 0.05:
30/06/15:
    - Passage de Mustache à Angular JS comme framework de template

## v. 0.04:
27/06/15:
    - Passage de Uikit à Tweeter Bootstrap comme framework de rendu (cf. templates/bootstrap.html)

## v. 0.03:
21/01/15:
    - Ajout de la possibilité d'utiliser le paramètre 'tpl' dans l'URL pour définir le template (ex.: tpl=content2.html)
    - Ajout de la possibilité d'accéder au document XML (en réalité le résultét du getRecordById)


js/mdReader.js
'use strict';
        
// Configuration de l'application
var app = {
    title: 'mdReaderJS',        // Titre de l'application
    name: 'mdReaderJS',         // Nom de l'application
    version: '0.05',            // Version de l'application
    template: 'bootstrap.html'  // Template par défaut de présentation de la fiche de métadonnées
};

// Configuration par défaut de l'url du CSW
var csw_config = {
    url: '',
    request: 'GetRecordById',
    service: 'CSW',
    version: '2.0.2',
    elementsetname: 'full',
    postencoding: 'XML',
    resulttype: 'results',
    outputschema: 'http://www.isotc211.org/2005/gmd',
    typenames: 'gmd:MD_Metadata',
    id: '',
    xml_dir: ''
};

// Définition de la langue par défaut pour les labels de la page.
var lang = 'fr';

// Lien vers le serveur de récupération des flux (gestion du pb de cross domain)
var server_url = './server/index.php';  // "false" pour un accès directe à la page sans passer par le script serveur si l'application cswReader est hébergée sur le serveur du flux CSW.

// Récupérer les valeurs passées en paramètres dans l'URL (GET)
var url_vars = getUrlParams(window.location);

// Ecraser la valeur du template avec le paramètre d'URL 'tpl'
if (url_vars['tpl']) {
    app.template = url_vars['tpl'];
}

// Modifier la configuration de l'url du CSW en fonction des valeurs passées en paramètre (GET)
csw_config.url = url_vars.url;
csw_config.id = url_vars.id;
csw_config.xml_dir = url_vars.xml_dir;

// Initialisation de la variable globale data
var data = {};
data['lb'] = lb[lang];
data['app'] = app;
data['csw_config'] = csw_config;

// Construction de l'URL du CSW    
var csw_url = urlConstruct(csw_config);
data['csw_url'] = csw_url;
console.log(csw_url);
if (server_url) {
    var url_page = server_url;
    var data_page = {url: csw_url};
} else {
    var url_page = csw_url;
    data_page = '';
}


/**
 * Déclaration de l'application Angular JS mdReaderApp
 */
var mdReaderApp = angular.module('mdReaderApp', [
    // Dépendance
    'mdReader'
]);

// Déclaration du module mdReader
var mdReader = angular.module('mdReader',[]);

// Contrôleur de l'application mdReader
mdReader.controller('mdReaderCtrl', ['$scope', '$http', function ($scope, $http) {
    $scope.xml = '';
    $scope.data = data;
    // console.log(url_page);
    // console.log(data_page);
    
    $http({
        method  : 'POST',
        url     : url_page,
        //url : '../../xml2csw/?request=GetRecordById&service=CSW&version=2.0.2&elementsetname=full&postencoding=XML&resulttype=results&outputschema=http://www.isotc211.org/2005/gmd&typenames=gmd:MD_Metadata&id=FR-236700019-ORTHO-RVB-20112012-CIGAL.xml&xml_dir=xml',
        dataType: 'json',
        data : data_page
    }).success(function(data, status, headers, config) {
        // console.log(data);  // XML document object
        $scope.data.md = readXML(data);
        // console.log($scope.data);
    }).error(function(data, status, headers, config) {
        console.log("Erreur: impossible de lire l'url demandée.");
        alert("Impossible de lire l'url demandée.");
    });
    
    // Définition du template utilisé pour l'affichage de la fiche de métadonnées
    $scope.template_url = "templates/"+app.template;
}]);
